dataset/bin.cpp

```diff
@@ -308,15 +351,86 @@ auto hide_masked(
   return make_non_owning_bins(zip(begin, end), buffer_dim, buffer);
 }
 
+template <class T> class TargetBins {
+public:
+  TargetBins(const VariableConstView &var) {
+    // In some cases all events in an input bin map to the same output, but
+    // right now bin<> cannot handle this and requires target bin indices for
+    // every bin element.
+    const auto &[begin_end, dim, buffer] = var.constituents<core::bin<T>>();
+    m_target_bins_buffer = makeVariable<scipp::index>(buffer.dims());
+    m_target_bins = make_non_owning_bins(begin_end, dim,
+                                         VariableView(m_target_bins_buffer));
+  }
+  auto &operator*() noexcept { return m_target_bins; }
+
+private:
+  Variable m_target_bins_buffer;
+  Variable m_target_bins;
+};
+
 } // namespace
 
+/// Reduce a dimension by concatenating bin contents of all bins along a
+/// dimension.
+///
+/// This is used to implement `concatenate(var, dim)`.
+template <class T>
+Variable concat_bins(const VariableConstView &var, const Dim dim) {
+  TargetBinBuilder builder;
+  builder.erase(dim);
+  TargetBins<T> target_bins(var);
+  builder.build(*target_bins, std::map<Dim, Variable>{});
+  auto [buffer, bin_sizes] = bin<DataArray>(var, *target_bins, builder.dims());
+  squeeze(bin_sizes, {dim});
+  const auto end = cumsum(bin_sizes);
+  const auto buffer_dim = buffer.dims().inner();
+  return make_bins(zip(end - bin_sizes, end), buffer_dim, std::move(buffer));
+}
+template Variable concat_bins<Variable>(const VariableConstView &, const Dim);
```

Comment:
Why are these needed?

Reply:
Essentially `concatenate` uses the same mechanism as `bin`, with the added twist that it may erase dimensions. Earlier in this branch I had extended `bin` (lower down) to support this, but I found the code hard to read. Splitting into three different top-level functions made this more easily understandable.

Reply:
I guess my question was more why do you need the `Variable` and `DataArray` template specifications here

Reply:
Those are *explicit template instantiations*. They allow putting the template declaration in the header, but the implementation in the source file.